### DISCUSSION TITLE
Ability to add callbacks to custom fields that can return data to the frontend Alpine component

### DISCUSSION THREAD

**Discussion by darrena092** [filamentphp/filament#5644] [Ideas]
Hi all 👋

I've been working on a custom field for a project and found it necessary to be able to pass some data back and forth between the UI and my custom field class.

Since we extend most of the Filament resource pages anyways, for various reasons unrelated to this, I was able to add in a "field actions" system, where field classes can register callbacks that both receive data from and return data to the frontend Alpine component.

It was kind of inspired by the built-in select component, as it has a few functions in a trait which is used by the pages to support searching.

Would this be something that could be useful for others? I'm more than happy to port this into Filament and submit a PR - I just want to see what the general feeling is about it before I spend the time doing it.

Thanks! 🙂

**Comment by danharrin** [comment]
Hey! Good idea but we do have a couple of similar features. Repeaters and Builders use "events" to execute callbacks (although they don't return anything). And we also have an existing "actions" system that is able to open modals as well. We use it for the `createOptionForm()` for `Select` fields. In v3 we're going to refactor the Repeater and Builder to use actions.

Is there anything your actions system does that ours doesn't?

**Reply by darrena092** [reply]
I'm not super familiar with your actions system, though I have seen the use of events in the builder and concluded that it wasn't suitable for what I was trying to do for the same reason that you mention - I couldn't return data from it.

Is it possible to return data from your Actions? I'm not familiar with them.

For an example of what I've added does. It's something like this in the custom field class:

```php
$this->registerFieldAction('getSearchResults', function ($query) {
    // Use $query (which is a search string) to find relevant data and return array.
})
```

That can then be provided to the Alpine component like this:

```html
<div x-data="myCustomComponent({
            state: $wire.entangle('{{ $getStatePath() }}').defer,
            getSearchResults: async (query) => {
                return $wire.invokeFieldAction('{{ $getStatePath() }}', 'getSearchResults', query);
            }
     })"
        id="{{ $getId() }}"
    >
</div>
```

So I guess it just provides a really simple way to keep these sorts of functions which provide data that previously needed to exist in the Livewire component / page as a part of the field class itself.

**Reply by danharrin** [reply]
https://github.com/filamentphp/filament/commit/6566cb419074987dea1f28d64d376b1cfa316be9

Try this

```php
->registerActions([
    'returnTheState' => fn ($state) => $state,
])
```

```blade
<div x-data="{ state: $wire.entangle('{{ $getStatePath() }}').defer }">
    <input x-model="state" type="text" />

    <button type="button" x-on:click="console.log(await $wire.mountFormComponentAction('{{ $getStatePath() }}', 'dd'))">
        console.log()
    </button>
</div>
```

**Reply by danharrin** [reply]
it'll be available in the next v2 release btw, but for now you can use a local copy of Filament to test if you'd like
